### DISCUSSION TITLE
Relations without foreign keys?

### DISCUSSION THREAD

**Discussion by davesteinberg** [prisma/prisma#7274] [Q&A] [answered]
I'm very new to Prisma, so I'm not sure if I'm misunderstanding or just plain missing something, but it seems to me that using relations to define connections between models always creates a foreign key to enforce referential integrity. Is this true?

If so, this limitation seems to be preventing me from modeling the data for the scenario that I'm trying to implement. Here's what I would like to do. I hope someone can point out if this is currently possible and, if so, how.

Basically, I want to be able to do a left outer join and to have that reflected in my generated types. Here's an abbreviated form of my example:

```prisma
model Story {
  id         String      @id @default(cuid())
  createdAt  DateTime    @default(now())
  content    String
  code       String
  postalCode PostalCode? @relation(fields: [code], references: [code])
}

model PostalCode {
  code     String   @id
  province Province
  name     String
}

enum Province { ... }
```

The idea here is that users create stories and, when they do, they must enter the first 3 digits of their postal code. They do not provide any further information about the postal code. Instead, we have preloaded the database with information about a whole bunch of postal codes (again, really just the first 3 digits).

Later, when we query the stories to display them, we want to be able to retrieve the corresponding postal code data, so that if they enter R8A, we'll be able to display that the story comes from Flin Flon, MB. If they enter a postal code that we don't have data for (because it's a newly created postal code and our data isn't up to date yet, or even just because they entered a fake postal code), that's totally fine. We can get back null for the postalCode and we'll just display the code they entered.

In this scenario, I definitely don't want the referential integrity constraints that I would get with a foreign key. I just want an efficient way to enhance the data with additional data, if it exists. Is it possible, or should it be?

I could imagine adding some other argument to the `@relation` attribute to indicate that I'm not interested in the reverse relation and I don't want referential integrity, but I don't see any sign that that's possible today.

Thanks in advance for your help.

**Comment by ryands17** [comment]
@davesteinberg 👋 

So if I understand correctly, if a user enters a postal code that exists, only the stories pertaining to the code should be shown. And if a postal code that doesn't exist is returned, then all stories should be retuened?

**Reply by davesteinberg** [reply]
@ryands17 Not quite, it's a bit simpler than that. Sorry, I was unclear about who does what when.

One user enters their postal code when they submit their story.

Later, when other users view the stories, they don't enter anything. They just see a list of stories. For each one, if we have data about the story's postal code, we will show the viewer the placename where that story came from. If we don't, they'll just see the postal code that the story submitter had entered.

**Reply by ryands17** [reply]
I don't think you need to create relations without foreign keys for this. The stories that do not have a postal code can be easily displayed via:

```ts
const stories = await prisma.story.findMany({
  include: { postalCode: true }
})
```

If the code is present, it will be shown else it wont. So you'll be fine keeping the constraints as well. You can check out the docs [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries) too.

**Reply by davesteinberg** [reply]
But the model I have above defines a foreign key that forbids me from having a story with a code that doesn't match an existing postal code. It errors if I even try to add one. That's why I'm looking for something similar but without the foreign key?

**Comment by janpio** [comment, marked as answer]
Not the solution to your problem, but to state the facts: 
Yes, if you define a relation in your Prisma Schema, Prisma Migrate will create a foreign key for that relation. If you introspect and existing database, Introspection will require foreign keys to identify relations.

You can add _additional_ relations to your Prisma schema that are not backed by a foreign key, but you are responsible for maintaining referential integrity in these cases which can be quite cumbersome.

**Reply by davesteinberg** [reply]
Thanks, I really appreciate the information.

I'm not quite sure what you're referring to when you say "add additional relations to your schema that are not backed by a foreign key." Could you please elaborate?

Do you mean my Prisma schema or my database schema? If Prisma, how would I do that? If database, how would I use those when querying with Prisma?

**Reply by janpio** [reply]
Sorry, edited to say "Prisma schema". You can just add `@relation` in your schema if you also have the other bits you need (a relation scalar field for `fields`, and a field it `references`). Then Prisma does not really care if there is a real _foreign key_ defined that matches that relation or not.

**Reply by davesteinberg** [reply]
Thanks again, but I'm still not getting it. What is the difference between these two things you're describing?

1. "Yes, if you define a relation in your Prisma Schema, Prisma Migrate will create a foreign key for that relation."
2. "You can just add @relation in your schema if you also have the other bits you need (a relation scalar field for fields, and a field it references)."

What exactly do you do differently to make Prisma create a foreign key or not create a foreign key?

**Reply by janpio** [reply]
In one case you run Prisma Migrate to migrate your database with a foreign key.
In the other case you create your database manually, run `prisma introspect` and then add the relation manually to have the Client API that lets you work with this relation.

But as said originally, I do not think this will actually help with your problem.

**Reply by davesteinberg** [reply]
Oh, I see! Out of curiosity, would it also work if I initially started with a Prisma schema without the `postalCode` relation, used `prisma migrate`, then added the relation, and ran `primsa generate` without creating a new migration?

**Reply by janpio** [reply]
Yes, that could indeed lead to the same end result.

But if adding the relation manually works, then really adding it to the database will also work. This is really just useful for databases with FK support or other good reasons.

**Reply by davesteinberg** [reply]
@janpio I just gave it a try and it works! Fantastic! Thank you so much!

There's just one problem: If I ever want to evolve my models and run `prisma migrate` again, I'll need to temporarily remove the `postalCode` relation while I do.

Would you consider adding more explicit support for this scenario? It would just require a way for me to tell Prisma that I don't actually want a foreign key to enforce the relation. In my original question, I suggested another, optional argument to the `@relation` attribute. But any mechanism would be good by me. Then, you could ignore the relation in `prisma migrate`, and it would be painless for me to keep using it in the future.

**Reply by janpio** [reply]
Perhaps - if we understand the use case.

Right now we are only aware of a binary choice: A database either supports and uses foreign keys, or they do not. Right now Prisma only supports databases that use foreign keys for relations. Next we will implement a mode for databases without foreign keys, like e.g. planetscale.com or MySQL MyISAM.

Can you elaborate why you want to have foreign keys on some, and not on others?

**Reply by davesteinberg** [reply]
For sure, I'd be happy to try to help you understand the use case.

I tried to write it up in my original question, but I was a bit ambiguous about which user is doing what, so I'll give it another go. Briefly, it's about using some relevant data, if available, to enrich another entity. The key thing is "if available". It's a weaker relationship in that you don't know if that relevant data will be there until you look, as opposed to a definite relationship between two entities that should/can be enforced.

My schema already has regular relationships with the need for referential integrity, but now I want to add one of these weaker relationships, as well.

This is the relevant portion of my model. I showed an almost identical version above, in my original question:

```
model Story {
  id         String      @id @default(cuid())
  createdAt  DateTime    @default(now())
  content    String
  code       String
  postalCode PostalCode? @relation(fields: [code], references: [code])
}

model PostalCode {
  code     String   @id
  province Province
  name     String
  hotspot  Boolean
}

enum Province { ... }
```

The `code` in `Story` is what defines the weak relationship between stories and postal codes. Every story will always have a `code`, but there may or may not be an actual `PostalCode` entity with that code.

I'm going to be initializing the `PostalCode` table as part of deploying the app, loading it up with data for each postal code, including an official placename from the post office, along with information from other sources, such as whether the postal code is a hotspot. Users won't ever change or add to this data. It will only be updated when I receive new data from the post office, process it, and load it.

When a user submits a story, they must type in the digits of their postal code. It may match one in the `PostalCode` table or it may not. It might not match one if it's a newly created postal code and our data isn't up to date yet or even just because they entered a fake postal code. That's totally fine.

Later, when another user visits the site, we'll retrieve the list of stories to display to them. At that point, I want to join that with the postal code data by the code that the submitter had entered. That way, I'll be able to show the story enhanced with additional information about the postal code: the placename and an indicator if it's a hotspot. If there is no matching `PostalCode` entity, no problem, I'll just show the code the submitter typed, with no name and no hotspot indicator.

Clearly this kind of query is totally possible without having a foreign key. Thanks to your suggestion, it's already working! The only thing a foreign key would do is prevent adding the story in the first place if it doesn't match a postal code. That's why a foreign key won't work in this situation.

And there's one extra complication: we already have the `Story` model (without the `postalCode` relation) in the application, which is live and running today. Currently, we just display the postal code that the submitter typed. I want to add the `PostalCode` model and data so that we can show more helpful information about the stories we have and the new stories to come. With hundreds of stories and no postal codes in the database, I can't even get as far as successfully running migrate on the schema if doing so adds the foreign key.

**Reply by davesteinberg** [reply]
Thanks for the thumbs up! I've created [issue 7351](https://github.com/prisma/prisma/issues/7351) to track the request.

**Reply by Zain-ul-din** [reply]
@janpio please can you answer this [question.](https://github.com/prisma/prisma/discussions/17012)

**Comment by rorhug** [comment]
@davesteinberg Is there a better solution to this yet other than adding/removing the @relation between migrations?

I guess one option is to just not use a relation, but then you lose the type-safety and "dsl" of joining.

Another one [is this](https://github.com/prisma/prisma/issues/7351#issuecomment-1326639397).

**Reply by davesteinberg** [reply]
@rorhug I'm not aware of a better solution currently. Not using a relation doesn't seem like a reasonable option, at least for my use case probably in general, since I guess it would mean doing joins manually.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
